### PR TITLE
hdf4: modernize, fix cross

### PR DESCRIPTION
**Pull request by TomaSajt**
- format
- use `hash`
- use `finalAttrs`
- use `--replace-fail`
- use `lib.cmake*` utility functions
  - the logic was changed a bit: if a toggle is disabled, the cmake flag is set to false instead of being unset.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by knedlsepp — APPROVED** [review]
Thanks for the cleanup!

**Comment by knedlsepp** [comment]
@NixOS/nixpkgs-merge-bot merge

**Comment by TomaSajt** [comment]
> @knedlsepp merge not permitted (#305350): 
> pkgs/tools/misc/hdf4/default.nix is not in pkgs/by-name/
> 

Other than being in pkgs/by-name, IIRC the PR has to be made by r-ryantm (the auto-update bot) to be mergeable with the merge bot.

File: pkgs/tools/misc/hdf4/default.nix [hunk 1]
```diff
@@ -1,99 +1,114 @@
...
 
-  cmakeFlags = [
-    "-DBUILD_SHARED_LIBS=ON"
-    "-DHDF4_BUILD_TOOLS=ON"
-    "-DHDF4_BUILD_UTILS=ON"
-    "-DHDF4_BUILD_WITH_INSTALL_NAME=OFF"
-    "-DHDF4_ENABLE_JPEG_LIB_SUPPORT=ON"
-    "-DHDF4_ENABLE_NETCDF=${if netcdfSupport then "ON" else "OFF"}"
-    "-DHDF4_ENABLE_Z_LIB_SUPPORT=ON"
-    "-DJPEG_DIR=${libjpeg}"
```
Comment (TomaSajt) [inline pkgs/tools/misc/hdf4/default.nix:83]: It turns out `JPEG_DIR` no longer does anything, so I removed it. Maybe the library detection got better.

**Comment by TomaSajt** [comment]
I also included and improved the cross-compilation fixes from https://github.com/NixOS/nixpkgs/pull/348319

**Comment by GaetanLepage** [comment]
[This commit](https://github.com/NixOS/nixpkgs/commit/6f6d588e1f416282cbfee5ed0c7f389822d8b3db) has broken `python3Packages.gdal` on all platforms.

**Comment by GaetanLepage** [comment]
I would strongly encourage you to run `nixpkgs-review` before merging PRs as much as possible. I know that it is not always feasible. Here it would have prevented a somewhat unnecessary breakage of several packages.

**Comment by TomaSajt** [comment]
Sorry about this...
Opened https://github.com/NixOS/nixpkgs/pull/353979

**Comment by GaetanLepage** [comment]
> Sorry about this...
> Opened #353979

No worry, thanks for the quick reaction :)

**Pull request merged**